File: Day158_2025/Q158_GFG.cpp
class Solution {
  public:
    int longestCommonSum(vector<int> &a1, vector<int> &a2) {
        // Q. Longest Span in two Binary Arrays
        // Code here.
        int n = a1.size();
        int res = 0;
        vector<int> diff(2 * n + 1, -1);
        int sum1 = 0, sum2 = 0;
        for (int i = 0; i < n; i++) {
            sum1 += a1[i];
            sum2 += a2[i];
            int currentDiff = sum1 - sum2;
            int index = currentDiff + n;
            if (currentDiff == 0) {
                res = max(res, i + 1);
            }
            else if (diff[index] != -1) {
                res = max(res, i - diff[index]);
            } 
            else {
                diff[index] = i;
            }
        }
        
        return res;
    }
};

File: Day158_2025/Q158_GFG.py
class Solution:
    def longestCommonSum(self, a1, a2):
        n = len(a1)
        diff = [-1] * (2 * n + 1)
        sum1 = sum2 = res = 0

        for i in range(n):
            sum1 += a1[i]
            sum2 += a2[i]
            currentDiff = sum1 - sum2
            index = currentDiff + n

            if currentDiff == 0:
                res = max(res, i + 1)
            elif diff[index] != -1:
                res = max(res, i - diff[index])
            else:
                diff[index] = i

        return res
